crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1897 @@
+
+            match block.node {
+                Node::Leaf(..) => return Some((item.index, block)),
+                Node::Internal(ref node) => {
+                    if item.visited {
+                        return Some((item.index, block));
+                    };
+
+                    self.deque.push_front(MerkleBlobLeftChildFirstIteratorItem {
+                        visited: true,
+                        index: item.index,
+                    });
+                    self.deque.push_front(MerkleBlobLeftChildFirstIteratorItem {
+                        visited: false,
+                        index: node.right,
+                    });
+                    self.deque.push_front(MerkleBlobLeftChildFirstIteratorItem {
+                        visited: false,
+                        index: node.left,
+                    });
+                }
+            }
+        }
+    }
+}
+
+pub struct MerkleBlobParentFirstIterator<'a> {
+    blob: &'a [u8],
+    deque: VecDeque<TreeIndex>,
+}
+
+impl<'a> MerkleBlobParentFirstIterator<'a> {
+    fn new(blob: &'a [u8]) -> Self {
+        let mut deque = VecDeque::new();
+        if blob.len() / BLOCK_SIZE > 0 {
+            deque.push_back(TreeIndex(0));
+        }
+
+        Self { blob, deque }
+    }
+}
+
+impl Iterator for MerkleBlobParentFirstIterator<'_> {
+    type Item = (TreeIndex, Block);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // left sibling first, parents before children
+
+        let index = self.deque.pop_front()?;
+        let block_bytes: BlockBytes = self.blob[block_range(index)].try_into().unwrap();
+        let block = Block::from_bytes(block_bytes).unwrap();
+
+        if let Node::Internal(ref node) = block.node {
+            self.deque.push_back(node.left);
+            self.deque.push_back(node.right);
+        }
+
+        Some((index, block))
+    }
+}
+
+pub struct MerkleBlobBreadthFirstIterator<'a> {
+    blob: &'a [u8],
+    deque: VecDeque<TreeIndex>,
+}
+
+impl<'a> MerkleBlobBreadthFirstIterator<'a> {
+    #[allow(unused)]
+    fn new(blob: &'a [u8]) -> Self {
+        let mut deque = VecDeque::new();
+        if blob.len() / BLOCK_SIZE > 0 {
+            deque.push_back(TreeIndex(0));
+        }
+
+        Self { blob, deque }
+    }
+}
+
+impl Iterator for MerkleBlobBreadthFirstIterator<'_> {
+    type Item = Block;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // left sibling first, parent depth before child depth
+
+        loop {
+            let index = self.deque.pop_front()?;
+            let block_bytes: BlockBytes = self.blob[block_range(index)].try_into().unwrap();
+            let block = Block::from_bytes(block_bytes).unwrap();
+
+            match block.node {
+                Node::Leaf(..) => return Some(block),
+                Node::Internal(node) => {
+                    self.deque.push_back(node.left);
+                    self.deque.push_back(node.right);
+                }
+            }
+        }
+    }
+}
```

Comment:
it doesn't look like these iterators have very thorough tests

Reply:
looks good. if it breaks in the future, it doesn't look like CI would tell us though